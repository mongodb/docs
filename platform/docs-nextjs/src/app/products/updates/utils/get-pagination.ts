/**
 *
 * @param array - flatten array
 * @param chunkSize - max size of each array
 * @returns [[{}, {}], [{}]]
 * @description Takes an flatten array and creates one level nested array of objects based on the chunkSize. Used for pagination data.
 */
export const getPagination = <T>(array: T[], chunkSize: number): T[][] => {
  const chunks: T[][] = [];
  if (!array.length) {
    return [[]]; // default to one empty nested array
  }
  for (let i = 0; i < array.length; i += chunkSize) {
    chunks.push(array.slice(i, i + chunkSize));
  }
  return chunks;
};
