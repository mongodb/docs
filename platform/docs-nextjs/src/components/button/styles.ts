import { css } from '@leafygreen-ui/emotion';
import { palette } from '@leafygreen-ui/palette';

const focusBoxShadow = (color: string) => `
  0 0 0 2px ${color}, 
  0 0 0 4px ${palette.blue.light1};
`;

// Theme-specific styles were copied from the original Button component
// to allow pre-rendering for dark mode
export const disabledStyle = css`
  &[aria-disabled='true'] {
    &,
    &:hover,
    &[data-hover='true'],
    &:active,
    &[data-active='true'] {
      background-color: ${palette.gray.light2};
      border-color: ${palette.gray.light1};
      color: ${palette.gray.base};
      box-shadow: none;
      cursor: not-allowed;
    }

    &:focus-visible,
    &[data-focus='true'] {
      color: ${palette.gray.base};
      box-shadow: ${focusBoxShadow(palette.white)};
    }

    // Needs to be nested here to ensure proper cascading
    .dark-theme & {
      &,
      &:hover,
      &[data-hover='true'],
      &:active,
      &[data-active='true'] {
        background-color: ${palette.gray.dark3};
        border-color: ${palette.gray.dark2};
        color: ${palette.gray.dark1};
        box-shadow: none;
        cursor: not-allowed;
      }

      &:focus-visible,
      &[data-focus='true'] {
        color: ${palette.gray.dark1};
        box-shadow: ${focusBoxShadow(palette.black)};
      }
    }
  }
`;
