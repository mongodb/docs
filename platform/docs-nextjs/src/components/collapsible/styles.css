/* minor css to change nth child properties in browsers without support for :nth-child(n of [selector]) */

.collapsible {
  border-top: 1px solid #e8edeb;
  margin-top: 32px;
}

.collapsible ~ .collapsible {
   border-top: none;
   margin-top: 0;
}