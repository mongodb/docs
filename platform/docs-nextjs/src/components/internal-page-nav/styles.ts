import { css } from '@leafygreen-ui/emotion';
import { palette } from '@leafygreen-ui/palette';

export const navLinkButtonStyle = css`
  background-color: ${palette.gray.light3};
  color: ${palette.black};

  .dark-theme & {
    background-color: ${palette.gray.dark2};
    color: ${palette.white};
  }
`;
