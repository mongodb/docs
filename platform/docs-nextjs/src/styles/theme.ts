const fontSize = {
  xsmall: '11px',
  tiny: '12px',
  small: '13px',
  default: '16px',
  h1: '36px',
  h2: '24px',
  h3: '18px',
  h4: '16px',
};

const size = {
  tiny: '4px',
  small: '8px',
  default: '16px',
  medium: '24px',
  large: '32px',
  xlarge: '64px',
  xxlarge: '128px',
  maxWidth: '1400px',
  stripUnit(unit: string): number {
    return parseInt(unit, 10);
  },
};

/**
 * store common responsive sizes as numbers
 */
const breakpoints = {
  xSmall: 320,
  small: 480,
  medium: 767,
  large: 1024,
  xLarge: 1200,
  xxLarge: 1440,
  xxxLarge: 1920,
};

/**
 * store common responsive sizes
 */
const screenSize = {
  upToXSmall: `only screen and (max-width: ${breakpoints.xSmall}px)`,
  xSmallAndUp: `not all and (max-width: ${breakpoints.xSmall}px)`,
  upToSmall: `only screen and (max-width: ${breakpoints.small}px)`,
  smallAndUp: `not all and (max-width: ${breakpoints.small}px)`,
  upToMedium: `only screen and (max-width: ${breakpoints.medium}px)`,
  mediumAndUp: `not all and (max-width: ${breakpoints.medium}px)`,
  upToLarge: `only screen and (max-width: ${breakpoints.large}px)`,
  largeAndUp: `not all and (max-width: ${breakpoints.large}px)`,
  upToXLarge: `only screen and (max-width: ${breakpoints.xLarge}px)`,
  xLargeAndUp: `not all and (max-width: ${breakpoints.xLarge}px)`,
  upTo2XLarge: `only screen and (max-width: ${breakpoints.xxLarge}px)`,
  '2XLargeAndUp': `not all and (max-width: ${breakpoints.xxLarge}px)`,
  upTo3XLarge: `only screen and (max-width: ${breakpoints.xxxLarge}px)`,
  '3XLargeAndUp': `not all and (max-width: ${breakpoints.xxxLarge}px)`,
  tablet: `only screen and (min-width: ${breakpoints.small + 1}px) and (max-width: ${breakpoints.large}px)`,
};

const header = {
  bannerHeight: '40px',
  navbarHeight: '95px',
  navbarMobileHeight: '56px',
  docsMobileMenuHeight: '52px',
  // used for scrolling elements into place, considering sticky header
  navbarScrollOffset: '85px',
  actionBarMobileHeight: '60px',
  composableDesktopHeight: '90px',
  composableMobileHeight: '190px',
};

const widgets = {
  buttonContainerMobileHeight: '60px',
};

const transitionSpeed = {
  iaExit: '100ms',
  iaEnter: '200ms',
  contentFade: '300ms',
};

// z-indexes for topmost major elements on the site. Subcomponents should have their own z-indexes set
const zIndexes = {
  content: 1,
  actionBar: 800,
  sidenav: 900,
  header: 1000,
  widgets: 2000,
  popovers: 10000,
};

// media queries that define system color preference
const colorPreference = {
  dark: '(prefers-color-scheme: dark)',
  light: '(prefers-color-scheme: light)',
  auto: '(prefers-color-scheme: auto)',
};

export const theme = {
  breakpoints,
  fontSize,
  header,
  widgets,
  screenSize,
  size,
  transitionSpeed,
  zIndexes,
  colorPreference,
};
