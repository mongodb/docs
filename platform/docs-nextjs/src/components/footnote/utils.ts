import type { ASTNode } from '@/types/ast';
import type { FootnoteNode, FootnoteReferenceNode } from '@/types/ast';
import { findAllKeyValuePairs } from '@/utils/find-all-key-value-pairs';
import type { FootnoteType } from './footnote-context';

export const getFootnotes = (nodes: ASTNode[]) => {
  const footnotes = findAllKeyValuePairs(nodes, 'type', 'footnote') as FootnoteNode[];
  const footnoteReferences = findAllKeyValuePairs(nodes, 'type', 'footnote_reference') as FootnoteReferenceNode[];
  const numAnonRefs = footnoteReferences.filter(
    (node) => !Object.prototype.hasOwnProperty.call(node, 'refname'),
  ).length;
  // We label our footnotes by their index, regardless of their names to
  // circumvent cases such as [[1], [#], [2], ...]
  return footnotes.reduce<Record<string, FootnoteType>>((map, footnote, index) => {
    if (footnote.name) {
      // Find references associated with a named footnote

      map[footnote.name] = {
        label: index + 1,
        references: getNamedFootnoteReferences(footnoteReferences, footnote.name),
      };
    } else {
      // Find references associated with an anonymous footnote
      // Replace potentially broken anonymous footnote ids
      footnote.id = `${index + 1}`;

      map[footnote.id] = {
        label: index + 1,
        references: getAnonymousFootnoteReferences(index, numAnonRefs),
      };
    }
    return map;
  }, {});
};

// Find all footnote_reference node IDs associated with a given footnote by
// that footnote's refname
const getNamedFootnoteReferences = (footnoteReferences: FootnoteReferenceNode[], refname: string) => {
  return footnoteReferences.filter((node) => node.refname === refname).map((node) => node.id);
};

// They are used infrequently, but here we match an anonymous footnote to its reference.
// The nth footnote on a page is associated with the nth reference on the page. Since
// anon footnotes and footnote references are anonymous, we assume a 1:1 pairing, and
// have no need to query nodes. If there are more anonymous footnotes than references,
// we may return an empty array
const getAnonymousFootnoteReferences = (index: number, numAnonRefs: number) => {
  return index > numAnonRefs ? [] : [`id${index + 1}`];
};
