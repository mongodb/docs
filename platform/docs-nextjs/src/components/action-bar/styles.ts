import { css } from '@leafygreen-ui/emotion';
import styled from '@emotion/styled';
import { theme } from '@/styles/theme';
import { palette } from '@leafygreen-ui/palette';
import { displayNone } from '@/utils/display-none';

export const ActionBarSearchContainer = styled.div`
  display: flex;
  align-items: center;
  width: 100%;
  background: inherit;

  @media ${theme.screenSize.mediumAndUp} {
    padding-right: ${theme.size.default};
  }

  @media ${theme.screenSize.upToLarge} {
    max-width: unset;
    justify-content: space-between;
    width: 100%;
  }
`;

interface StyledInputProps {
  sidenav?: boolean;
  mobileSearchActive: boolean;
}

export const StyledInputContainer = styled.div<StyledInputProps>`
  width: 100%;
  max-width: 610px;
  background: inherit;

  div[role='searchbox'] {
    background-color: var(--search-input-background-color);
    width: 100%;
  }

  --search-input-background-color: ${palette.white};
  .dark-theme & {
    --search-input-background-color: ${palette.gray.dark4};
  }

  @media ${theme.screenSize.mediumAndUp} {
    width: ${({ sidenav }) => (sidenav ? '70' : '100')}%;
  }

  ${(props) => {
    return (
      props.mobileSearchActive &&
      `
      display: flex !important;
      position: absolute;
      width: calc(100% - ${theme.size.medium} - ${theme.size.medium});
      z-index: 40;
      max-width: unset;
      left: ${theme.size.medium};
      column-gap: ${theme.size.medium};
      align-items: center;

      form {
        width: 100%;
        margin-right: ${theme.size.medium};
      }
    `
    );
  }}
`;

// Used to ensure dropdown is same width as input
export const StyledSearchBoxRef = styled.div`
  width: 100%;
`;

export const searchInputStyling = ({ mobileSearchActive }: StyledInputProps) => css`
  ${!mobileSearchActive && displayNone.onMedium};

  @media ${theme.screenSize.upToMedium} {
    input[type='search'] {
      font-size: ${theme.fontSize.default};
    }
  }

  ${mobileSearchActive &&
  `
    display: flex !important;
    width: calc(100% - ${theme.size.medium} - ${theme.size.medium});
    z-index: 40;
    max-width: unset;
    background: var(--background-color-primary);
    align-items: center;

    > form {
      width: 100%;
      margin-right: ${theme.size.medium};
    }
  `}
`;

export const ActionsBox = styled('div')`
  display: flex;
  align-items: center;
  column-gap: ${theme.size.default};
  position: relative;
  top: 0;
  margin: 0 ${theme.size.large} 0 ${theme.size.medium};
  justify-self: flex-end;
  grid-column: -2/-1;

  @media ${theme.screenSize.upToLarge} {
    column-gap: 6px;
    margin-right: ${theme.size.medium};
    margin-left: ${theme.size.small};
  }

  @media ${theme.screenSize.upToMedium} {
    margin-left: 1px;
  }
`;

export const overlineStyling = css`
  display: flex;
  flex-direction: row;
  align-items: center;
  color: ${palette.gray.dark1};
  text-transform: uppercase;
  text-wrap: nowrap;
  font-weight: 600;
  cursor: pointer;
  ${displayNone.onLargerThanTablet};
  font-size: ${theme.fontSize.tiny};
  padding-right: ${theme.size.large};
  .dark-theme & {
    color: ${palette.gray.light2};
  }

  > svg {
    margin-right: ${theme.size.small};
  }
`;

export const searchIconStyling = css`
  ${displayNone.onLargerThanMedium};
  float: right;
  justify-content: right;
`;

export const offlineStyling = css`
  @media ${theme.screenSize.largeAndUp} {
    display: none;
  }
`;

const hideOnEnLang = `
  &:not(:lang(EN)) {
    display: none;
  }
`;

export const chatbotButtonStyling = css`
  text-wrap-mode: nowrap;
  ${displayNone.onMobileAndTablet};
  ${hideOnEnLang}
`;

export const chatbotMobileButtonStyling = css`
  ${displayNone.onLargerThanTablet}
  ${hideOnEnLang}
  color: ${palette.green.dark2};

  .dark-theme & {
    color: ${palette.green.dark1};
  }
`;
