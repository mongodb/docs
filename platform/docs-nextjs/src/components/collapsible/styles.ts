import { css } from '@leafygreen-ui/emotion';
import { palette } from '@leafygreen-ui/palette';
import { theme } from '@/styles/theme';

export const collapsibleStyle = css`
  border-bottom: 1px solid ${palette.gray.light2};

  /* minor css to change nth child properties in browsers without support for :nth-child(n of [selector]) */

  & {
    border-top: 1px solid #e8edeb;
    margin-top: 32px;
  }

  & ~ & {
    border-top: none;
    margin-top: 0;
  }

  &:nth-child(1 of .collapsible) {
    border-top: 1px solid ${palette.gray.light2};
    margin-top: ${theme.size.large};
  }

  &:nth-last-child(1 of .collapsible) {
    margin-bottom: ${theme.size.large};
  }
`;

export const headerContainerStyle = css`
  display: flex;
  justify-content: space-between;
  column-gap: ${theme.size.medium};
  margin: ${theme.size.large} 0;
`;

export const headerStyle = css`
  margin-top: 0;
`;

export const iconStyle = css`
  align-self: center;
  flex-shrink: 0;
`;

export const innerContentStyle = css`
  overflow: hidden;
  height: 0;
  color: --font-color-primary;

  [aria-expanded='true'] & {
    height: auto;
  }
`;
