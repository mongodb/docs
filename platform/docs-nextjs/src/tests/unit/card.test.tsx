import { render } from '@testing-library/react';
import { ThemeProvider } from '@emotion/react';
import { mockLocation } from '@/tests/utils/mock-location';
import Card from '@/components/card';
import { theme } from '@/styles/theme';

// data for this component
import mockData from '@/tests/data/card.test.json';
import type { CardNode } from '@/types/ast';

// Mock the useRouter hook from next/navigation
const mockPush = jest.fn();
jest.mock('next/navigation', () => ({
  useRouter: () => ({
    push: mockPush,
  }),
}));

// Mock the PageContext
jest.mock('@/context/page-context', () => ({
  usePageContext: () => ({
    template: 'default',
  }),
}));

beforeAll(() => {
  mockLocation({ search: '', pathname: `/` });
});

it('renders correctly', () => {
  const cardNode = mockData as CardNode;
  const tree = render(
    <ThemeProvider theme={theme}>
      <Card nodeChildren={cardNode.children} {...cardNode.options} />
    </ThemeProvider>,
  );
  expect(tree.asFragment()).toMatchSnapshot();
});

describe('relative urls passed to Card', () => {
  test.each([
    ['/foo', 1],
    ['/foo/', 2],
    ['foo', 3],
  ])('should use router.push function to navigate to %s', (url, expectedCalls) => {
    const cardNode = mockData as CardNode;
    cardNode.options.url = url;
    const { getByText } = render(
      <ThemeProvider theme={theme}>
        <Card nodeChildren={cardNode.children} {...cardNode.options} isCompact={true} />
      </ThemeProvider>,
    );

    getByText('Test card headline').click();

    // Verify that router.push was called with the expected URL
    expect(mockPush).toHaveBeenCalledTimes(expectedCalls);
    expect(mockPush).toHaveBeenCalledWith(url);
  });
});
