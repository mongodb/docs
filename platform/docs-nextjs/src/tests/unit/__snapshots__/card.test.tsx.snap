// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
  -webkit-transition: 150ms ease-in-out;
  transition: 150ms ease-in-out;
  transition-property: border,box-shadow;
  border-radius: 24px;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  font-size: 13px;
  line-height: 20px;
  padding: 24px;
  min-height: 68px;
  border: 1px solid #E8EDEB;
  box-shadow: 0 4px 10px -4px rgba(0,30,43,0.3);
  background-color: #FFFFFF;
  color: #1C2D38;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  background-color: var(--background-color-primary);
  border-color: #E8EDEB;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 32px;
}

.emotion-0:focus {
  outline: none;
  box-shadow: 0 0 0 2px #FFFFFF,0 0 0 4px #0498EC,0 4px 10px -4px rgba(0,30,43,0.3);
}

.emotion-0:hover,
.emotion-0:active {
  border: 1px solid #E8EDEB;
  box-shadow: 0 4px 20px -4px rgba(0,30,43,0.2);
}

.emotion-0:hover:focus,
.emotion-0:active:focus {
  box-shadow: 0 0 0 2px #FFFFFF,0 0 0 4px #0498EC,0 4px 20px -4px rgba(0,30,43,0.2);
}

.dark-theme .emotion-0 {
  border-color: #3D4F58;
}

.emotion-0 p:last-of-type {
  margin-bottom: 0;
}

.emotion-1:last-of-type {
  margin-bottom: 16px;
}

.emotion-2 {
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 700;
  font-size: 12px;
  line-height: 16px;
  border-radius: 24px;
  height: 18px;
  padding-left: 6px;
  padding-right: 6px;
  text-transform: uppercase;
  border: 1px solid;
  letter-spacing: 0.4px;
  background-color: #E3FCF7;
  border-color: #C0FAE6;
  color: #00684A;
}

.emotion-3 {
  margin: unset;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #001E2B;
  font-size: 13px;
  line-height: 20px;
  color: #001E2B;
  font-weight: 500;
  font-weight: 500;
  letter-spacing: normal;
  color: var(--font-color-primary);
  margin: 16px 0 8px 0;
}

.emotion-3 strong,
.emotion-3 b {
  font-weight: 600;
}

.emotion-4 {
  margin: unset;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #001E2B;
  font-size: 13px;
  line-height: 20px;
  color: #001E2B;
  font-weight: 400;
  margin-bottom: 16px;
  color: var(--font-color-primary);
}

.emotion-4 strong,
.emotion-4 b {
  font-weight: 600;
}

.emotion-5 {
  margin: unset;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #001E2B;
  font-size: 13px;
  line-height: 20px;
  color: #001E2B;
  font-weight: 400;
}

.emotion-5 strong,
.emotion-5 b {
  font-weight: 600;
}

.emotion-5 a {
  line-height: unset;
}

.emotion-6 {
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  display: inline;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-decoration-color: transparent;
  cursor: pointer;
  font-size: inherit;
  line-height: inherit;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background: none;
  border: none;
  padding: 0;
  font-size: 13px;
  line-height: 20px;
  color: #016BF8;
  font-weight: 400;
  display: inline;
  color: var(--link-color-primary);
  font-weight: var(--link-font-weight);
}

.emotion-6:hover,
.emotion-6[data-hover='true'],
.emotion-6:focus-visible,
.emotion-6[data-focus='true'] {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  -webkit-transition: text-decoration 150ms ease-in-out;
  transition: text-decoration 150ms ease-in-out;
  text-underline-offset: 4px;
  text-decoration-thickness: 2px;
}

.emotion-6:focus {
  outline: none;
}

.emotion-6:hover,
.emotion-6[data-hover='true'] {
  text-decoration-color: #E8EDEB;
}

.emotion-6:focus-visible,
.emotion-6[data-focus='true'] {
  text-decoration-color: #016BF8;
}

.emotion-6>span>code,
.emotion-6>code {
  color: var(--link-color-primary);
  font-weight: var(--link-font-weight);
}

<div
    class="emotion-0"
  >
    <img
      alt="MongoDB University icon"
      class="
  @media only screen and (max-width: 480px) {
    width: 20px;
  }
"
      data-nimg="1"
      decoding="async"
      height="24"
      loading="lazy"
      src="/images/icons/university.svg"
      style="color: transparent;"
      width="24"
    />
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        Test card tag
      </div>
    </div>
    <div>
      <p
        class="emotion-3"
      >
        Test card headline
      </p>
      <p
        class="emotion-4"
      >
        Test card paragraph
      </p>
      <p
        class="emotion-5"
      >
        <a
          class="lg-ui-0001 emotion-6"
          href="https://university.mongodb.com/courses/M001/about"
          target="_self"
        >
          <span>
            Test card CTA
          </span>
        </a>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
